src/mtt_internals.c

```diff
@@ -103,8 +104,24 @@ MEMKIND_EXPORT void *mtt_internals_realloc(MttInternals *internals, void *ptr,
 {
     // TODO this is more complicated: realloc might call malloc, so we actually
     // need to extend pool
+
+    if (size == 0) {
+        mtt_internals_free(internals, ptr);
+        return NULL;
+    } else if (ptr == NULL) {
+        return mtt_internals_malloc(internals, size);
+    }
+
+    size_t old_size = mtt_internals_usable_size(internals, ptr);
+    if (size <= old_size) {
+        return ptr;
+    }
+
+    void *ret = mtt_internals_malloc(internals, size);
+    memcpy(ret, ptr, old_size);
     mtt_internals_free(internals, ptr);
```

Comment:
*IMPORTANT* the behavior of this function goes against the standard:

[If the function fails to allocate the requested block of memory, a null pointer is returned, and the memory block pointed to by argument ptr is not deallocated (it is still valid, and with its contents unchanged).](https://www.cplusplus.com/reference/cstdlib/realloc/)